File: src/scavenger/tui/widgets/splitter.py
"""Draggable splitter bars for resizing adjacent panes."""
from textual.widget import Widget
from textual.events import MouseDown, MouseMove, MouseUp


class VSplitter(Widget):
    """Vertical splitter — drag left/right to resize siblings."""

    can_focus = False

    def render(self) -> str:
        return ""

    DEFAULT_CSS = """
    VSplitter {
        width: 1;
        max-width: 1;
        height: 100%;
        background: #222;
    }
    VSplitter:hover {
        background: #555;
    }
    VSplitter.-dragging {
        background: #fd971f;
    }
    """

    def __init__(self, left_id: str, right_id: str, **kwargs) -> None:
        super().__init__(**kwargs)
        self._left_id = left_id
        self._right_id = right_id
        self._dragging = False
        self._start_x: int = 0
        self._initial_left: int = 0

    def on_mouse_down(self, event: MouseDown) -> None:
        try:
            left = self.screen.query_one(f"#{self._left_id}")
        except Exception:
            return
        self._dragging = True
        self._start_x = event.screen_x
        self._initial_left = left.size.width
        self.add_class("-dragging")
        self.capture_mouse()
        event.stop()
        event.prevent_default()

    def on_mouse_move(self, event: MouseMove) -> None:
        if not self._dragging:
            return
        event.stop()
        event.prevent_default()
        dx = event.screen_x - self._start_x
        if dx == 0:
            return
        new_left = self._initial_left + dx
        if new_left < 14:
            return
        try:
            left = self.screen.query_one(f"#{self._left_id}")
            right = self.screen.query_one(f"#{self._right_id}")
        except Exception:
            return
        remaining = self.parent.size.width - new_left - 1
        if remaining < 20:
            return
        left.styles.width = new_left
        right.styles.width = "1fr"

    def on_mouse_up(self, event: MouseUp) -> None:
        if self._dragging:
            self._dragging = False
            self.remove_class("-dragging")
            self.release_mouse()
            event.stop()


class HSplitter(Widget):
    """Horizontal splitter — drag up/down to resize siblings."""

    can_focus = False

    def render(self) -> str:
        return ""

    DEFAULT_CSS = """
    HSplitter {
        height: 1;
        max-height: 1;
        width: 100%;
        background: #222;
    }
    HSplitter:hover {
        background: #555;
    }
    HSplitter.-dragging {
        background: #fd971f;
    }
    """

    def __init__(self, top_id: str, bottom_id: str, **kwargs) -> None:
        super().__init__(**kwargs)
        self._top_id = top_id
        self._bottom_id = bottom_id
        self._dragging = False
        self._start_y: int = 0
        self._initial_bottom: int = 0

    def on_mouse_down(self, event: MouseDown) -> None:
        try:
            bottom = self.screen.query_one(f"#{self._bottom_id}")
        except Exception:
            return
        self._dragging = True
        self._start_y = event.screen_y
        self._initial_bottom = bottom.size.height
        self.add_class("-dragging")
        self.capture_mouse()
        event.stop()
        event.prevent_default()

    def on_mouse_move(self, event: MouseMove) -> None:
        if not self._dragging:
            return
        event.stop()
        event.prevent_default()
        dy = event.screen_y - self._start_y
        if dy == 0:
            return
        new_bottom = self._initial_bottom - dy
        if new_bottom < 3:
            return
        try:
            top = self.screen.query_one(f"#{self._top_id}")
            bottom = self.screen.query_one(f"#{self._bottom_id}")
        except Exception:
            return
        remaining = self.parent.size.height - new_bottom - 1
        if remaining < 5:
            return
        top.styles.height = "1fr"
        bottom.styles.height = new_bottom

    def on_mouse_up(self, event: MouseUp) -> None:
        if self._dragging:
            self._dragging = False
            self.remove_class("-dragging")
            self.release_mouse()
            event.stop()

File: rust/src/tui/widgets/splitter.rs
use ratatui::buffer::Buffer;
use ratatui::layout::Rect;
use ratatui::style::{Color, Style};
use ratatui::widgets::Widget;

use crate::tui::colors;

/// Vertical splitter state — drag left/right to resize sibling panes.
///
/// The splitter tracks whether a drag is active and the initial mouse/size
/// state at capture time. The parent layout reads `left_width` after each
/// mouse event to recompute the constraint split.
#[derive(Debug)]
pub struct VSplitterState {
    dragging: bool,
    hovering: bool,
    start_x: u16,
    initial_left: u16,
    /// Current width of the left pane. The parent owns the Rect split but
    /// reads this value to set constraints.
    pub left_width: u16,
    pub min_left: u16,
    pub min_right: u16,
}

impl VSplitterState {
    pub fn new(initial_left: u16, min_left: u16, min_right: u16) -> Self {
        Self {
            dragging: false,
            hovering: false,
            start_x: 0,
            initial_left,
            left_width: initial_left,
            min_left,
            min_right,
        }
    }

    pub fn is_dragging(&self) -> bool {
        self.dragging
    }

    pub fn on_mouse_down(&mut self, screen_x: u16, current_left_width: u16) {
        self.dragging = true;
        self.start_x = screen_x;
        self.initial_left = current_left_width;
    }

    /// Returns true if the left_width was updated.
    pub fn on_mouse_move(&mut self, screen_x: u16, parent_width: u16) -> bool {
        if !self.dragging {
            return false;
        }
        let dx = screen_x as i32 - self.start_x as i32;
        let new_left = (self.initial_left as i32 + dx).max(0) as u16;
        if new_left < self.min_left {
            return false;
        }
        // 1 column for the splitter bar itself
        let remaining = parent_width.saturating_sub(new_left + 1);
        if remaining < self.min_right {
            return false;
        }
        self.left_width = new_left;
        true
    }

    pub fn on_mouse_up(&mut self) {
        self.dragging = false;
    }

    pub fn set_hover(&mut self, hover: bool) {
        self.hovering = hover;
    }

    fn color(&self) -> Color {
        if self.dragging {
            colors::ORANGE
        } else if self.hovering {
            colors::SPLITTER_HOVER
        } else {
            colors::SPLITTER_IDLE
        }
    }
}

pub struct VSplitterWidget<'a> {
    state: &'a VSplitterState,
}

impl<'a> VSplitterWidget<'a> {
    pub fn new(state: &'a VSplitterState) -> Self {
        Self { state }
    }
}

impl Widget for VSplitterWidget<'_> {
    fn render(self, area: Rect, buf: &mut Buffer) {
        let style = Style::default().bg(self.state.color());
        for y in area.top()..area.bottom() {
            if let Some(cell) = buf.cell_mut((area.x, y)) {
                cell.set_style(style);
                cell.set_char(' ');
            }
        }
    }
}

/// Horizontal splitter state — drag up/down to resize sibling panes.
#[derive(Debug)]
pub struct HSplitterState {
    dragging: bool,
    hovering: bool,
    start_y: u16,
    initial_bottom: u16,
    /// Current height of the bottom pane.
    pub bottom_height: u16,
    pub min_top: u16,
    pub min_bottom: u16,
}

impl HSplitterState {
    pub fn new(initial_bottom: u16, min_top: u16, min_bottom: u16) -> Self {
        Self {
            dragging: false,
            hovering: false,
            start_y: 0,
            initial_bottom,
            bottom_height: initial_bottom,
            min_top,
            min_bottom,
        }
    }

    pub fn is_dragging(&self) -> bool {
        self.dragging
    }

    pub fn on_mouse_down(&mut self, screen_y: u16, current_bottom_height: u16) {
        self.dragging = true;
        self.start_y = screen_y;
        self.initial_bottom = current_bottom_height;
    }

    /// Returns true if the bottom_height was updated.
    pub fn on_mouse_move(&mut self, screen_y: u16, parent_height: u16) -> bool {
        if !self.dragging {
            return false;
        }
        let dy = screen_y as i32 - self.start_y as i32;
        let new_bottom = (self.initial_bottom as i32 - dy).max(0) as u16;
        if new_bottom < self.min_bottom {
            return false;
        }
        let remaining = parent_height.saturating_sub(new_bottom + 1);
        if remaining < self.min_top {
            return false;
        }
        self.bottom_height = new_bottom;
        true
    }

    pub fn on_mouse_up(&mut self) {
        self.dragging = false;
    }

    pub fn set_hover(&mut self, hover: bool) {
        self.hovering = hover;
    }

    fn color(&self) -> Color {
        if self.dragging {
            colors::ORANGE
        } else if self.hovering {
            colors::SPLITTER_HOVER
        } else {
            colors::SPLITTER_IDLE
        }
    }
}

pub struct HSplitterWidget<'a> {
    state: &'a HSplitterState,
}

impl<'a> HSplitterWidget<'a> {
    pub fn new(state: &'a HSplitterState) -> Self {
        Self { state }
    }
}

impl Widget for HSplitterWidget<'_> {
    fn render(self, area: Rect, buf: &mut Buffer) {
        let style = Style::default().bg(self.state.color());
        for x in area.left()..area.right() {
            if let Some(cell) = buf.cell_mut((x, area.y)) {
                cell.set_style(style);
                cell.set_char(' ');
            }
        }
    }
}
